python/srf/_pysrf/src/segment.cpp

```diff
@@ -181,22 +200,37 @@ std::shared_ptr<srf::segment::ObjectProperties> SegmentProxy::make_sink(srf::seg
     return self.make_sink<PyHolder, PythonSink>(name, on_next_w, on_error_w, on_completed_w);
 }
 
-std::shared_ptr<srf::segment::ObjectProperties> SegmentProxy::get_ingress(
-    srf::segment::Builder& self,
-    const std::string& name)
+std::shared_ptr<srf::segment::ObjectProperties> SegmentProxy::get_ingress(srf::segment::Builder& self,
+                                                                          const std::string& name)
 {
+    auto it_caster = node::PortRegistry::s_port_to_type_index.find(name);
+    if (it_caster != node::PortRegistry::s_port_to_type_index.end())
+    {
+        VLOG(2) << "Found an ingress port caster for " << name;
+
+        return self.get_ingress_dynamic(name, it_caster->second);
+    }
     return self.get_ingress<PyHolder>(name);
 }
 
-std::shared_ptr<srf::segment::ObjectProperties> SegmentProxy::get_egress(
-    srf::segment::Builder& self,
-    const std::string& name)
+std::shared_ptr<srf::segment::ObjectProperties> SegmentProxy::get_egress(srf::segment::Builder& self,
+                                                                         const std::string& name)
 {
-   return self.get_egress<PyHolder>(name);
+    auto it_caster = node::PortRegistry::s_port_to_type_index.find(name);
```

Comment:
Why is this doing a lookup in a static list of port names to types? There may be a time when we have multiple ports with the same name but different types (like a derived/base class situation).

At this point, the segment proxy should be able to lookup the ingress/egress port object by name and query the type id from there.

Reply:
> Why is this doing a lookup in a static list of port names to types? There may be a time when we have multiple ports with the same name but different types (like a derived/base class situation).

Ryan said this is not allowed by SRF (as in we do not allow ports of the same name and different data types) and not something we need to worry about.

> At this point, the segment proxy should be able to lookup the ingress/egress port object by name and query the type id from there.

I don't believe this is correct given the way things work. The current implementation of `Builder::get_ingress` first calls `get_ingress_base`; which returns an `IngressPortBase`; and is able to up cast it because it has the type information. Absent this information we can't directly cast an IngressPortBase to an ObjectProperties object that we can call `sink_source()` on, because ObjectProperties is a virtual base of the object. To resolve this, I went with storing a name->type_index map.

Reply:
We have discussed this and it will require a larger change to make creating manifolds more explicit